mmf/trainers/base_trainer.py

```diff
@@ -278,16 +280,49 @@ def _forward_pass(self, batch):
         return report
 
     def _backward(self, loss):
-        self.optimizer.zero_grad()
+        # for update_frequency > 1, scale the loss so that accumulated gradients are the
+        # expected magnitude
+        loss = loss / self.config.training.update_frequency
         loss.backward()
+        self.profile("Backward time")
+
+    def _check_start_update(self, epoch_batch_index, epoch_num_batches):
+        # When epoch_num_batches is not a multiple of the effective batch size, the last
+        # few batches
+        # will be discarded. I.e., they won't contribute to an update.
```

Comment:
Even after discarding, the forward and backward is still happening without `optimizer.step()` and `optimizer.zero_grad()`. Do we really need to skip them? what issues arise when you don't skip them?

Reply:
I understand it is hard to keep track of how many iterations among `update_frequency` iterations were completed. But maybe it is better to keep track of those explicitly rather than blindly dividing with `update_frequency`?

Reply:
We're talking about restructuring the loop in a separate comment and offline chat. I'll revisit these comments after we've made a decision on the loop structure.